fix: build spec experience

AlexD10S opened the pull request:
Fixes https://github.com/r0gue-io/pop-cli/issues/305

### Changes

1. Add the flag `--chain` to specify the chain specification. Is to add this flag: https://github.com/paritytech/polkadot-sdk/blob/master/substrate/client/cli/src/params/shared_params.rs#L32 when running the `build-spec` command of the binary.
To generate `pop-node`  chain_spec mainnet for example can be done with:
`pop build spec --release --id 4024 --type live --relay paseo  --chain mainnet`
 (https://github.com/r0gue-io/pop-node/blob/main/node/src/command.rs#L73)

2. As pointed in the issue the `--default-bootnode` was a bit confusing because by default was set to true and to don't  include localhost as the default bootnode has to be specified with  `--default-bootnode false`.
Set by default to false and included if specified `--default-bootnode`.
The other approach is to change the name to `--disable-default-bootnode` and set it by default to false unless the flag is included.

brunopgalvao commented:
Looking good. Some notes:

`--release` option is confusing, what does this do? It says: "For production, always build in release mode to exclude debug features", so will the `pop build spec --release` build the parachain in release mode? 

I think `--release` should be removed and the default should be so that when the user runs `pop build spec` it will build in release mode by default.

```
pop build spec --release --id 4024 --type live --relay paseo --chain mainnet 
```
For some reason this does not work, not sure if it is a pop-cli issue or a pop-node issue, also something to keep in mind, `mainnet` on pop-node uses the polkadot relay, so I guess the user would have to be aware to use the appropriate flags to generate the chain spec.

AlexD10S commented:
> Looking good. Some notes:
> 
> `--release` option is confusing, what does this do? It says: "For production, always build in release mode to exclude debug features", so will the `pop build spec --release` build the parachain in release mode?
> 
> I think `--release` should be removed and the default should be so that when the user runs `pop build spec` it will build in release mode by default.
> 
> ```
> pop build spec --release --id 4024 --type live --relay paseo --chain mainnet 
> ```
> 
> For some reason this does not work, not sure if it is a pop-cli issue or a pop-node issue, also something to keep in mind, `mainnet` on pop-node uses the polkadot relay, so I guess the user would have to be aware to use the appropriate flags to generate the chain spec.

You are right about the `--release` flag been confusing, similar as the problem above with `--default-bootnode` flag. I like your suggestion, I deprecated it  the `release` flag and it builds in RELEASE mode by default.

About this: `pop build spec --release --id 4024 --type live --relay paseo --chain mainnet ` not working, you probably have an old version of pop-node. See https://github.com/r0gue-io/pop-node/blob/main/node/src/command.rs#L73
You can check if your build has the mainnet available running:` ./target/release/pop-node build-spec --chain mainnet `

evilrobot-01 commented on crates/pop-cli/src/commands/build/spec.rs:
```diff
@@ -134,18 +134,22 @@ pub struct BuildSpecCommand {
 	#[arg(short = 'o', long = "output")]
 	pub(crate) output_file: Option<PathBuf>,
-	/// For production, always build in release mode to exclude debug features.
+	/// [DEPRECATED] For production, always build in release mode to exclude debug features.
```
Why is this deprecated? There is nothing in the PR description and no comment here to advise why this is being done.

evilrobot-01 commented on crates/pop-cli/src/commands/build/spec.rs:
```diff
@@ -134,18 +134,22 @@ pub struct BuildSpecCommand {
 	pub(crate) id: Option<u32>,
 	/// Whether to keep localhost as a bootnode.
-	#[clap(long, default_value = "true")]
+	#[clap(long, default_value = "false")]
```
```suggestion
	#[clap(long)]
```
Surely a default value for a bool is false?

evilrobot-01 commented on crates/pop-cli/src/commands/build/spec.rs:
```diff
@@ -195,7 +199,9 @@ impl BuildSpecCommand {
 		if !self.release {
-			cli.warning("NOTE: this command defaults to DEBUG builds for development chain types. Please use `--release` (or simply `-r` for a release build...)")?;
+			cli.warning(
+				"NOTE: this flag is deprecated, this command defaults to a release build builds.",
```
```suggestion
				"NOTE: this flag is deprecated, this command defaults to a release build.",
```

evilrobot-01 replied:
Some context in a comment as to why this is deprecated would be appreciated please.

evilrobot-01 reviewed (commented):
Looks good, would just appreciate some context in terms of the deprecation of release option. Why not leave the flexibility?

Would also want @al3mart to look this PR over before being merged.

AlexD10S commented:
> Looks good, would just appreciate some context in terms of the deprecation of release option. Why not leave the flexibility?
> 
> Would also want @al3mart to look this PR over before being merged.

My apologies for not mentioning it in the PR description. The deprecation of the `--release` flag was implemented in response to a feedback comment https://github.com/r0gue-io/pop-cli/pull/331#issuecomment-2456156539 on the flag being confusing.

The changes were made in a single [commit 37bb25d](https://github.com/r0gue-io/pop-cli/pull/331/commits/37bb25d5ff2aadd774e64614dc90f34b89c6c86a). so will be easy to revert if we decide to keep it. What are your thoughts, @al3mart? A better approach suggested?

al3mart commented on crates/pop-cli/src/commands/build/spec.rs:
```diff
@@ -134,18 +134,22 @@ pub struct BuildSpecCommand {
+	/// Specify the chain specification. It can be one of the predefined ones (dev, local) or a
+	/// custom one.
+	#[arg(short = 'c', long = "chain", default_value = "dev")]
+	pub(crate) chain: Option<String>,
```
<3

al3mart reviewed (commented):
build spec might trigger a compilation. I feel that having the option of declaring if we want the profile to be `release` or `debug` can be convenient .

If we didn't have the flag anywhere else I think removing it might make more sense.
The flag was added mainly so the options for builds were consistent across sub commands
Maybe we can improve the doc as a way to clarify why that flag is present..

AlexD10S commented:
> build spec might trigger a compilation. I feel that having the option of declaring if we want the profile to be `release` or `debug` can be convenient .
> 
> If we didn't have the flag anywhere else I think removing it might make more sense. The flag was added mainly so the options for builds were consistent across sub commands Maybe we can improve the doc as a way to clarify why that flag is present..

You are right, It makes sense to keep it consistent across the comment. I removed the deprecation message and added more information in the docs to clarify that this command builds the node if it hasn't been built yet.

brunopgalvao commented:
> build spec might trigger a compilation. I feel that having the option of declaring if we want the profile to be `release` or `debug` can be convenient .
> 
> If we didn't have the flag anywhere else I think removing it might make more sense. The flag was added mainly so the options for builds were consistent across sub commands Maybe we can improve the doc as a way to clarify why that flag is present..

What is the difference in the outputted chain spec artifacts when using debug vs release?

brunopgalvao commented:
> About this: `pop build spec --release --id 4024 --type live --relay paseo --chain mainnet ` not working, you probably have an old version of pop-node. See https://github.com/r0gue-io/pop-node/blob/main/node/src/command.rs#L73
> You can check if your build has the mainnet available running:`./target/release/pop-node build-spec --chain mainnet`

Yes you are right. Since there are quite a few options for this command that are needed, it would be nice to have an error if for example `--chain mainnet` does not exist instead of the CLI doing nothing which is what it currently does.

AlexD10S commented:
> Yes you are right. Since there are quite a few options for this command that are needed, it would be nice to have an error if for example `--chain mainnet` does not exist instead of the CLI doing nothing which is what it currently does.

Yes, makes sense. Logs of the command `./target/release/pop-node build-spec` are hidden, maybe we should show them if there is an error: https://github.com/r0gue-io/pop-cli/issues/337

al3mart commented:
> > build spec might trigger a compilation. I feel that having the option of declaring if we want the profile to be `release` or `debug` can be convenient .
> > If we didn't have the flag anywhere else I think removing it might make more sense. The flag was added mainly so the options for builds were consistent across sub commands Maybe we can improve the doc as a way to clarify why that flag is present..
> 
> What is the difference in the outputted chain spec artifacts when using debug vs release? Can the runtime wasm be different?

Regarding the resulting wasm I'd say that the only possible difference is that if it is not built in release it may contain some extra debug information that can be logged. Release also tries to compile with optimization. This comes to mind - https://paritytech.github.io/polkadot-sdk/master/frame_support/derive.RuntimeDebugNoBound.html

evilrobot-01 commented on crates/pop-cli/src/commands/build/spec.rs:
```diff
@@ -195,13 +200,14 @@ impl BuildSpecCommand {
 
-		let spinner = cliclack::spinner();
-		spinner.start("Generating chain specification...");
+		cli.info("Generating chain specification...")?;
```
Why all the info outputs instead of the spinner? I'd imagine its just going to flood the terminal with messages which will just be annoying after the first time.

AlexD10S replied:
It was to fix the previous logs: https://github.com/r0gue-io/pop-cli/pull/334

al3mart reviewed (approved):
I think the change on default value for `--default-bootnode` makes sense. I can see how that was confusing.

Love the addition of `--chain`. Was definitely something needed.

evilrobot-01 commented on crates/pop-cli/src/commands/build/spec.rs:
```diff
@@ -206,8 +206,7 @@ impl BuildSpecCommand {
 		}
-
-		cli.info("Generating chain specification...")?;
+		let multi = multi_progress("Generating chain specification...");
```
Not sure multi has much value here - there are no actions happening in parallel which need to be shown at the same time and the output in the spinner is replaced in any case. Would simplify the code to just use a single spinner.

AlexD10S replied:
Solved! It looks nicer now with the `spinner`. I had to include a `spinner.clear();` in case the node has to be built otherwise output in the terminal very messy as Ale indicated here https://github.com/r0gue-io/pop-cli/pull/334 (That was the reason to use `multi_progress`).

evilrobot-01 commented on crates/pop-cli/src/commands/build/spec.rs:
```diff
@@ -284,23 +285,25 @@ impl BuildSpecCommand {
 		}
 
-		cli.intro("Building your chain spec".to_string())?;
+		spinner.stop("Done.");
```
If using a single spinner, would be better as 'Chain specification built successfully.' or something.

evilrobot-01 reviewed (changes requested):
I think the multi/spinner can be improved a little quite easily.

evilrobot-01 commented:
As mentioned separately, there is no consideration for builds using the production profile, which is potentially an important use case. Whilst this may have no effect on the resulting wasm, it did mean a rebuild of the node despite me already having a production build at target/production. This obviously is a waste of time and local storage and also might be frustrating to an end user.

This might be an edge case, but I think the tooling should ideally cater to both novice and professional devs.

al3mart reviewed (approved):
LGTM.

Looking forward to merging this one!

AlexD10S commented on crates/pop-cli/src/commands/build/spec.rs:
```diff
@@ -127,169 +131,393 @@ pub(crate) enum RelayChain {
 	pub(crate) output_file: Option<PathBuf>,
-	/// For production, always build in release mode to exclude debug features.
-	#[clap(short = 'r', long, default_value = "true")]
+	// TODO: remove at release ...
```
🫡

Daanvdplas replied:
Added the upcoming release

AlexD10S commented on crates/pop-cli/src/commands/build/spec.rs:
```diff
@@ -299,156 +527,370 @@ impl BuildSpecCommand {
 	}
+
+	// Customize a chain specification.
+	fn customize(&self) -> anyhow::Result<()> {
```
Just fix the 2 warnings here and we are good to merge

Daanvdplas replied:
Thank you, resolved

al3mart reviewed (approved):
The new commits after my last approval seem to make sense. Thanks for addressing those!
